### PR TITLE
[huggingface_tensorflow, huggingface_pytorch][build] Pass arguments into dockerfile at buildtime

### PR DESCRIPTION
**Pull request by SergTogul**
*Issue #, if available:*

## PR Checklist
- [ ] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [ ] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [ ] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

## Pytest Marker Checklist
- [ ] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [ ] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [ ] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [ ] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron">)` to the new tests which I have added, if a test is specifically applicable to only one processor type

#### EIA/NEURON Checklist
* When creating a PR:
- [ ] I've modified `src/config/build_config.py` in my PR branch by setting `ENABLE_EI_MODE = True` or `ENABLE_NEURON_MODE = True`
* When PR is reviewed and ready to be merged:
- [ ] I've reverted the code change on the config file mentioned above
#### Benchmark Checklist
* When creating a PR:
- [ ] I've modified `src/config/test_config.py` in my PR branch by setting `ENABLE_BENCHMARK_DEV_MODE = True`
* When PR is reviewed and ready to be merged:
- [ ] I've reverted the code change on the config file mentioned above

## Reviewer Checklist
* For reviewer, before merging, please cross-check:
- [ ] I've verified the code change on the config file mentioned above has already been reverted

*Description:*

*Tests run:*

*DLC image/dockerfile:*

*Additional context:*

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/image_builder.py [hunk 1]
```diff
@@ -110,6 +110,16 @@ def image_builder(buildspec):
                 labels[var] = file_name
                 labels[f"{var}_URI"] = uri
 
+        if str(BUILDSPEC["framework"]).startswith("huggingface"):
+            if image_config.get("transformers_version"):
```
Comment (saimidu) [inline src/image_builder.py:114]: QQ: Is there a way to implement a "membership-check" for the key, like `if "transformers_version" in image_config`?

This is generally done for python classes through https://docs.python.org/3/reference/datamodel.html#object.__contains__

File: src/image_builder.py [hunk 2]
```diff
@@ -110,6 +110,16 @@ def image_builder(buildspec):
...
+                extra_build_args["TRANSFORMERS_VERSION"] = image_config.get("transformers_version")
+            else:
+                raise Exception(f"HuggingFace buildspec.yml must contain 'transformer_version' field for each image")
```
Comment (saimidu) [inline src/image_builder.py:117]: Please make this a specific Exception like `MissingImageBuildParameterException`, or generic like a `KeyError`.

File: huggingface/pytorch/training/docker/1.6/py3/cu110/Dockerfile.gpu [hunk 3]
```diff
@@ -4,11 +4,15 @@ LABEL maintainer="Amazon AI"
...
 
 # install Hugging Face libraries and its dependencies
 RUN pip install --no-cache-dir \ 
 	transformers[sklearn,sentencepiece]==${TRANSFORMERS_VERSION} \ 
 	datasets==${DATASETS_VERSION}
-RUN apt-get update && apt install git-lfs
+RUN apt-get update && \
+    apt install git-lfs && \
+    apt-get clean && \
+    rm -rf /…/apt-cache
```
Comment (saimidu) [inline huggingface/pytorch/training/docker/1.6/py3/cu110/Dockerfile.gpu:17]: Change to:
```suggestion
RUN apt-get update \
 && apt install git-lfs \
 && apt-get clean \
 && rm -rf /var/lib/apt/lists/*
```

File: huggingface/pytorch/training/docker/1.7/py3/cu110/Dockerfile.gpu [hunk 4]
```diff
@@ -8,11 +8,14 @@ LABEL maintainer="Amazon AI"
...
 
 # install Hugging Face libraries and its dependencies
 RUN pip install --no-cache-dir \ 
 	transformers[sklearn,sentencepiece]==${TRANSFORMERS_VERSION} \ 
 	datasets==${DATASETS_VERSION}
-RUN apt-get update && apt install git-lfs
+RUN apt-get update && \
+    apt install git-lfs && \
+    apt-get clean && \
+    rm -rf /…/apt-cache
```
Comment (saimidu) [inline huggingface/pytorch/training/docker/1.7/py3/cu110/Dockerfile.gpu:21]: Change to:
```suggestion
RUN apt-get update \
 && apt install git-lfs \
 && apt-get clean \
 && rm -rf /var/lib/apt/lists/*
```

File: huggingface/tensorflow/training/docker/2.3/py3/cu110/Dockerfile.gpu [hunk 5]
```diff
@@ -4,12 +4,16 @@ LABEL maintainer="Amazon AI"
...
 RUN pip install --no-cache-dir \
     transformers[sklearn,sentencepiece]==${TRANSFORMERS_VERSION} \
     datasets \
     tensorflow-addons==0.12.0 \
     psutil
-RUN apt-get update && apt install git-lfs
+RUN apt-get update && \
+    apt install git-lfs && \
+    apt-get clean && \
+    rm -rf /…/apt-cache
```
Comment (saimidu) [inline huggingface/tensorflow/training/docker/2.3/py3/cu110/Dockerfile.gpu:19]: Change to:
```suggestion
RUN apt-get update \
 && apt install git-lfs \
 && apt-get clean \
 && rm -rf /var/lib/apt/lists/*
```

File: huggingface/tensorflow/training/docker/2.4/py3/cu110/Dockerfile.gpu [hunk 6]
```diff
@@ -8,13 +8,16 @@ LABEL maintainer="Amazon AI"
...
 RUN pip install --no-cache-dir \
     transformers[sklearn,sentencepiece]==${TRANSFORMERS_VERSION} \
     datasets==${DATASETS_VERSION} \
     tensorflow-addons==0.12.0 \
     psutil
-RUN apt-get update && apt install git-lfs
+RUN apt-get update && \
+    apt install git-lfs && \
+    apt-get clean && \
+    rm -rf /…/apt-cache
```
Comment (saimidu) [inline huggingface/tensorflow/training/docker/2.4/py3/cu110/Dockerfile.gpu:23]: Change to:
```suggestion
RUN apt-get update \
 && apt install git-lfs \
 && apt-get clean \
 && rm -rf /var/lib/apt/lists/*
```

File: src/image_builder.py [hunk 7]
```diff
@@ -110,6 +110,16 @@ def image_builder(buildspec):
                 labels[var] = file_name
                 labels[f"{var}_URI"] = uri
 
+        if str(BUILDSPEC["framework"]).startswith("huggingface"):
+            if "transformers_version" in image_config:
```
Comment (saimidu) [inline src/image_builder.py:114]: Is `image_config` definitely a dict, or is it an object of a custom class?

File: src/image_builder.py [hunk 8]
```diff
@@ -110,6 +110,16 @@ def image_builder(buildspec):
...
+                extra_build_args["TRANSFORMERS_VERSION"] = image_config.get("transformers_version")
+            else:
+                raise KeyError(f"HuggingFace buildspec.yml must contain 'transformer_version' field for each image")
```
Comment (saimidu) [inline src/image_builder.py:117]: Nice.

However, the field we're looking for is transformer**s**_version.

**Review by saimidu — COMMENTED** [review]
Please also address dockerfile change requests.

File: src/image_builder.py (hunk 7 above)
Comment (SergTogul) [reply src/image_builder.py:114]: It's a CommentedMap rather. But it also supports __contains__()

**Review by saimidu — APPROVED** [review]

**Pull request merged**